src/Cms/Binding.php

```diff
@@ -0,0 +1,417 @@
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+class Binding
+{
+	/**
+	 * Local collection for more performant
+	 * limited index to lookup model
+	 */
+	protected Models|null $collection;
+
+	/**
+	 * Model instance
+	 */
+	protected Site|User|Page|File|null $model;
+
+	/**
+	 * UUID as protocol instance
+	 */
+	protected BindingProtocol $protocol;
+
+
+	public function __construct(
+		string|null $uuid = null,
+		Site|User|Page|File|null $model = null,
+		Models|null $collection = null
+	) {
+		$this->model      = $model;
+		$this->protocol   = BindingProtocol::for($uuid ?? $model);
+		$this->collection = $collection;
+	}
+
+	/**
+	 * Get instance for lookup cache
+	 *
+	 * @return \Kirby\Cache\Cache
+	 */
+	public static function cache(): Cache
+	{
+		return App::instance()->cache('uuid');
+	}
+
+	/**
+	 * Returns the index collection to consider
+	 * when traveersing content files
+	 *
+	 * @param bool $global
+	 * @return \Kirby\Cms\Models
+	 */
+	protected function collection(bool $global = false): Models
+	{
+		// prefer current collection as lookup context
+		// to potentially speed up finding the matching model
+		if ($this->collection !== null && $global === false) {
+			return $this->collection;
+		}
+
+		$kirby = App::instance();
+
+		// gather the collection depending on model type
+		$site       = $kirby->site();
+		$collection = match ($this->type()) {
+			'page'  => $site->index(true),
+			'file'  => $site->index(true)->files()
+				->add($site->files())
+				->add($kirby->users()->files()),
+			// TODO: indexes for content fields
+			// 'block'     => null,
+			// 'structure' => $site->index(true)->content()
+		};
+
+		// if a current collection is set, we know this is
+		// already the second time we call ::findFromIndex,
+		// which is why we can remove this current collection from
+		// the global index as we have already checked it
+		if ($this->collection !== null) {
+			$collection = $collection->remove($this->collection);
+		}
+
+		return $collection;
+	}
+
+	/**
+	 * Clears UUID from cache
+	 *
+	 * @param bool $recursive also clear subpages from cache
+	 * @return bool
+	 */
+	public function clear(bool $recursive = false): bool
+	{
+		// For pages: if $recursive, also clear UUIDs
+		// from cache for all children recursively
+		if ($this->type() == 'page' && $recursive === true) {
+			foreach ($this->model->children() as $child) {
+				static::for($child)->clear(true);
+			}
+		}
+
+		return static::cache()->remove($this->key());
+	}
+
+	/**
+	 * Create a new UUID hash and update
+	 * the model's content file to include it
+	 *
+	 * @return string
+	 */
+	public function create(): void
+	{
+		$id = Str::uuid();
+
+		// TODO: updating block/structure would work different
+
+		// update content file with generated id:
+		// make sure Kirby has the required permissions
+		// for the update action
+		$kirby = App::instance();
+		$user  = $kirby->auth()->currentUserFromImpersonation();
```

Comment:
Can't we just use …

```suggestion
		$user  = $kirby->user();
```

Reply:
Not sure if that's always the same, e.g. when code around already is impersonating anyone. It's exactly what `$kirby->impersonate()` does. In 3.9 we will be bale to streamline this bit of code, once the callback for `$kirby->impersonate()` doesn't bind the app instance anymore.

Reply:
Ah, you are right. You could be impersonating someone and still have another logged in user. Right. I got you. Then we need to keep it like this until 3.9